geopandas/io/tests/test_file.py

```diff
@@ -316,3 +316,91 @@ def test_read_file_empty_shapefile(tmpdir):
     empty = read_file(fname)
     assert isinstance(empty, geopandas.GeoDataFrame)
     assert all(empty.columns == ["A", "Z", "geometry"])
+
+
+def test_write_index_to_file(tmpdir, df_points):
+    tempfilename = os.path.join(str(tmpdir), "check.shp")
+
+    def do_checks(df, index_is_used):
+        # check combinations of index=None|True|False on GeoDataFrame/GeoSeries
+        other_cols = list(df.columns)
+        other_cols.remove("geometry")
+
+        index_cols = list(df.index.names)
+        if not index_is_used:
+            # pandas' default index name if not provided is 'index'
+            index_cols = ["index"]
+
+        # check GeoDataFrame with default index=None to autodetect
+        df.to_file(tempfilename, index=None)
+        df_check = read_file(tempfilename)
+        expected_cols = []
+        if len(other_cols) == 0:
+            # note: ESRI Shapefile will add FID if no other columns are used
+            expected_cols.append("FID")
+        if index_is_used:
+            expected_cols += index_cols
+        expected_cols += other_cols + ["geometry"]
+        assert list(df_check.columns) == expected_cols
+
+        # similar check on GeoSeries with index=None
+        df.geometry.to_file(tempfilename, index=None)
+        df_check = read_file(tempfilename)
+        if index_is_used:
+            expected_cols = index_cols + ["geometry"]
+        else:
+            expected_cols = ["FID", "geometry"]
+        assert list(df_check.columns) == expected_cols
+
+        # check GeoDataFrame with index=True
+        df.to_file(tempfilename, index=True)
+        df_check = read_file(tempfilename)
+        assert list(df_check.columns) == index_cols + other_cols + ["geometry"]
+
+        # similar check on GeoSeries with index=True
+        df.geometry.to_file(tempfilename, index=True)
+        df_check = read_file(tempfilename)
+        assert list(df_check.columns) == index_cols + ["geometry"]
+
+        # check GeoDataFrame with index=False
+        df.to_file(tempfilename, index=False)
+        df_check = read_file(tempfilename)
+        if len(other_cols) == 0:
+            expected_cols = ["FID", "geometry"]
+        else:
+            expected_cols = other_cols + ["geometry"]
+        assert list(df_check.columns) == expected_cols
+
+        # similar check on GeoSeries with index=False
+        df.geometry.to_file(tempfilename, index=False)
+        df_check = read_file(tempfilename)
+        assert list(df_check.columns) == ["FID", "geometry"]
+
+    # default range index
+    df_p = df_points.copy()
+    df = GeoDataFrame(df_p["value1"], geometry=df_p.geometry)
+    do_checks(df, False)
+
+    # index is different than a default range index
+    df_p = df_points.copy()
+    df_p.index += 1
```

Comment:
Small additional comment: I just checked and doing `+= 1` still keeps it as a range index (but not starting at 0, which is probably good to test as well). 
But I think it would be good to also test something else as a range index (in the case of no name), eg a Int64Index (for that, you could do eg `df.index = list(range(1, len(df) + 1))`, that should give a int index).

Reply:
Good idea, I've add this scenario, and a scenario where one row is dropped, which also creates an Int64Index.